tsl/src/continuous_aggs/invalidation.c

```diff
@@ -0,0 +1,389 @@
+		values[AttrNumberGetAttrOffset(
+			Anum_continuous_aggs_materialization_invalidation_log_modification_time)] =
+			Int64GetDatum(inval_form->modification_time);
+		values[AttrNumberGetAttrOffset(
+			Anum_continuous_aggs_materialization_invalidation_log_lowest_modified_value)] =
+			Int64GetDatum(inval_form->lowest_modified_value);
+		values[AttrNumberGetAttrOffset(
+			Anum_continuous_aggs_materialization_invalidation_log_greatest_modified_value)] =
+			Int64GetDatum(inval_form->greatest_modified_value);
+
+		/*  Create one log entry per aggregate and invalidation */
+		foreach (lc, cagg_ids)
+		{
+			int32 cagg_hyper_id = lfirst_int(lc);
+
+			values[AttrNumberGetAttrOffset(
+				Anum_continuous_aggs_materialization_invalidation_log_materialization_id)] =
+				Int32GetDatum(cagg_hyper_id);
+
+			ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+			ts_catalog_insert_values(state->cagg_log_rel, tupdesc, values, nulls);
+			ts_catalog_restore_user(&sec_ctx);
+		}
+
+		ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+		ts_catalog_delete(ti_inval->scanrel, inval_tuple);
+		ts_catalog_restore_user(&sec_ctx);
+
+		if (should_free)
+			heap_freetuple(inval_tuple);
+
+		MemoryContextSwitchTo(oldmctx);
+		MemoryContextReset(state->per_tuple_mctx);
+	}
+
+	ts_scan_iterator_close(&iterator);
+}
+
+static HeapTuple
+create_invalidation_tup(const HeapTuple tuple, int64 start, int64 end)
+{
+	HeapTuple newtup = heap_copytuple(tuple);
+	Form_continuous_aggs_materialization_invalidation_log form =
+		(Form_continuous_aggs_materialization_invalidation_log) GETSTRUCT(newtup);
+
+	form->lowest_modified_value = start;
+	form->greatest_modified_value = end;
+
+	return newtup;
+}
+
+static void
+cagg_invalidations_scan_by_hypertable_init(ScanIterator *iterator, int32 cagg_hyper_id,
+										   LOCKMODE lockmode)
+{
+	*iterator = ts_scan_iterator_create(CONTINUOUS_AGGS_MATERIALIZATION_INVALIDATION_LOG,
+										lockmode,
+										CurrentMemoryContext);
+	iterator->ctx.index = catalog_get_index(ts_catalog_get(),
+											CONTINUOUS_AGGS_MATERIALIZATION_INVALIDATION_LOG,
+											CONTINUOUS_AGGS_MATERIALIZATION_INVALIDATION_LOG_IDX);
+	ts_scan_iterator_scan_key_init(
+		iterator,
+		Anum_continuous_aggs_materialization_invalidation_log_materialization_id,
+		BTEqualStrategyNumber,
+		F_INT4EQ,
+		Int32GetDatum(cagg_hyper_id));
+}
+
+/*
+ * Process all cagg invalidations that match a refresh window.
+ *
+ * An invalidation entry is either completely enclosed (covered) by the
+ * refresh window, or it partially overlaps. In the former case, the
+ * invalidation entry is removed and for the latter case it is cut. Thus, an
+ * entry can either disappear, reduce in size, or be cut in two.
+ *
+ * Note that the refresh window is inclusive at the start and exclusive at the
+ * end.
+ */
+static void
+process_cagg_invalidations_for_refresh(CaggInvalidationState *state,
```

Comment:
I don't understand the purpose of the function. It just removes invalidations inside the window. How are cagg data updated inside the window on the invalidation? Is it entire window refreshed always independently of the window? Then the name of the function and comment should reflect this.

Reply:
After a refresh, all the cagg data in the window is up-to-date, so there should be no invalidations in that window. 

Currently, this function only clears the invalidations, it doesn't use them to optimize the refresh since the current refresh is brute-force (i.e., the whole window is recomputed). But that doesn't change the fact that we need to clear the invalidations. 

I can change the name to `clear_invalidations_for_refresh` for now if that makes more sense to you.

Reply:
It will be great to improve the name and the comment, so it corresponds to current implementation. I was confused by `process` and it took more time than necessary to understand what it is going on there.

Reply:
I changed the name, and clarified in comment and commit message.